astacus/coordinator/plugins/base.py

```diff
@@ -275,6 +291,123 @@ async def run_step(self, cluster: Cluster, context: StepsContext) -> List[ipc.No
+        if self.retention.minimum_backups is not None and self.retention.minimum_backups >= len(kept_backups):
+            return kept_backups
+        now = utils.now()
+
+        manifests = [await download_backup_manifest(self.json_storage, backup) for backup in kept_backups]
+        manifests = sorted(manifests, key=lambda m: (m.start, m.end, m.filename))
+        while manifests:
+            if self.retention.maximum_backups is not None:
+                if self.retention.maximum_backups < len(manifests):
+                    manifests.pop()
+                    continue
+
+            # Ok, so now we have at most <maximum_backups> (if set) backups
+            # Do we have too _few_ backups to delete any more?
+            if self.retention.minimum_backups is not None:
+                if self.retention.minimum_backups >= len(manifests):
+                    break
+
+            if self.retention.keep_days is not None:
+                manifest = manifests[-1]
+                if (now - manifest.end).days > self.retention.keep_days:
+                    manifests.pop()
+                    continue
+            # We don't have any other criteria to filter the backup manifests with
+            break
+
+        return set(manifest.filename for manifest in manifests)
+
+
+@dataclasses.dataclass
+class DeleteBackupManifestsStep(Step[set[str]]):
+    """
+    Delete all backup manifests that are not kept.
+    """
+
+    json_storage: AsyncJsonStorage
+
+    async def run_step(self, cluster: Cluster, context: StepsContext) -> set[str]:
+        all_backups = context.get_result(ListBackupsStep)
+        kept_backups = context.get_result(ComputeKeptBackupsStep)
+        deleted_backups = all_backups - kept_backups
+        for backup in deleted_backups:
+            logger.info("deleting backup manifest %r", backup)
+            await self.json_storage.delete_json(backup)
+        return deleted_backups
+
+
+@dataclasses.dataclass
+class DownloadKeptBackupManifestsStep(Step[Sequence[ipc.BackupManifest]]):
+    """
+    Download the manifest of all kept backups.
+    """
+
+    json_storage: AsyncJsonStorage
+
+    async def run_step(self, cluster: Cluster, context: StepsContext) -> Sequence[ipc.BackupManifest]:
+        backup_names = context.get_result(ComputeKeptBackupsStep)
+        return [await download_backup_manifest(self.json_storage, backup_name) for backup_name in backup_names]
+
+
+@dataclasses.dataclass
+class DeleteDanglingHexdigestsStep(Step[None]):
+    """
+    Delete all backups that are not kept.
+    """
+
+    hexdigest_storage: AsyncHexDigestStorage
+
+    async def run_step(self, cluster: Cluster, context: StepsContext) -> None:
+        deleted_backups = context.get_result(DeleteBackupManifestsStep)
+        if not deleted_backups:
+            logger.debug("no backup deleted, skipping delete dangling hexdigests")
+            return
+        kept_manifests = context.get_result(DownloadKeptBackupManifestsStep)
```

Comment:
my non-blocking 2c:

So ```deleted_backups``` is used only for skipping, not for actual deletion. If we get some misalignment between the DeleteBackupManifestsStep and DownloadKeptBackupManifestsStep , we could introduce some bug.

Of course, the name of the class is correct, but then, if we delete all dangling hexdigests, we could that every time, no skipping if there're no deleted backup. Or, maybe there's a reason for this and some comment would be welcome

Reply:
I followed the logic of the existing code before splitting it.

It's an optimization to minimize unnecessary costs when repeatedly calling cleanup and there is nothing to do: if no backups were recently deleted we don't incur the cost of downloading manifests (although these are cahed) or listing all existing hexdigests (these are not cached).

But looking at the surrounding code, we only call cleanup when there is a backup to delete, so the optimization is not worth the increase in complexity, I'll remove it.